PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,449 @@
+            if module_name in self:
+                del self[module_name]
+
+
+#FIXME: For space efficiency, convert to use slots rather than "__dict__".
+class ModuleHook(object):
+    """
+    Cached object encapsulating a lazy loadable hook script.
+
+    This object's `code` property provides the currently loaded contents (e.g.,
+    attributes, functions) of this hook script. The first access of this
+    property silently lazily loads this script into an in-memory code object.
+
+    Parameters
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    module_name : str
+        Name of the module hooked by this hook script.
+    hook_filename : str
+        Absolute or relative path of this hook script.
+    hook_module_name : str
+        Name of the in-memory module of this hook script's code object.
+    _code : code
+        Code object of the interpreted contents of this hook script, lazily
+        loaded on the first access of the `code` property _or_ `None` if this
+        property has yet to be accessed.
+    """
+
+    ## Private methods
+
+    def __init__(
+        self,
+        module_graph,
+        module_name,
+        hook_filename,
+        hook_module_name_prefix,
+    ):
+        """
+        Initialize this metadata.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        module_name : str
+            Name of the module hooked by this hook script.
+        hook_filename : str
+            Absolute or relative path of this hook script.
+        hook_module_name_prefix : str
+            String prefixing the name of the in-memory module for this hook
+            script's code object. To avoid namespace clashes with similar
+            modules created by other `ModuleHook` objects in other
+            `ModuleHookCache` containers, this string _must_ be unique to the
+            `ModuleHookCache` container containing this `ModuleHook` object. If
+            this string is non-unique, an existing in-memory module will be
+            erroneously reused when lazily loading this hook script, thus
+            erroneously resanitizing previously sanitized hook script attributes
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
+        self.hook_module_name = (
+            hook_module_name_prefix + self.module_name.replace('.', '_'))
+
+        # Attributes defined by lazily loading this hook.
+        self._code = None
+
+    ## Properties
+
+    @property
+    def code(self):
```

Comment:
Because this isn't actually a code object, and other parts of pyinstaller (and modulegraph) do deal with code objects in various ways, I'd rename it to something like `module` or `namespace` since it is really the namespace of the hook file's module that is being accessed. 

<sub>The hook file does have a code object, but it is executed somewhere inside the import function using the module's namespace as its globals dict.</sub>